___

<a href='https://github.com/pick1'> <img src='tensorflow.jpg' /></a>
___

# Tensorflow Project 
This Deep Learning project looked at the effectiveness of Neural Nets!

It used the [Iris Data Set](https://archive.ics.uci.edu/ml/datasets/iris) from the UCI repository.

The data consists of 3 classes of irises and 4 columns containing:

1. sepal length in cm 
2. sepal width in cm 
3. petal length in cm 
4. petal width in cm 
5. class: 
-- Iris Setosa 
-- Iris Versicolour 
-- Iris Virginica

Where class indicates which family of iris the sample belonged to.

This is a perfect task for Neural Networks and Deep Learning!

## Loading Libraries and Dataset

In [1]:
from sklearn.datasets import load_iris
import tensorflow as tf
iris = load_iris()

/usr/local/lib/python3.5/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [2]:
iris.keys()

dict_keys(['target', 'feature_names', 'DESCR', 'data', 'target_names'])

## Setting X and y variables

In [3]:
X = iris['data']
y = iris['target']

## Splitting data into train and test sets

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

## Loading the Sci-kit Learn TensorFlow Library

In [5]:
import tensorflow.contrib.learn as learn

## Creating a classifier

In [6]:
feature_columns = [tf.contrib.layers.real_valued_column("",dimension = 3)]
classifier = learn.DNNClassifier(hidden_units=[10, 20, 10], n_classes = 3, feature_columns = feature_columns)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_evaluation_master': '', '_session_config': None, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_task_id': 0, '_save_summary_steps': 100, '_keep_checkpoint_max': 5, '_tf_config': gpu_options {
  per_process_gpu_memory_fraction: 1.0
}
, '_log_step_count_steps': 100, '_keep_checkpoint_every_n_hours': 10000, '_tf_random_seed': None, '_master': '', '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7133b02d8fd0>, '_num_ps_replicas': 0, '_environment': 'local', '_num_worker_replicas': 0, '_is_chief': True, '_task_type': None, '_model_dir': '/tmp/tmp2ywxwlmv'}


## Fitting the data to the Classifier

In [8]:
classifier.fit(X_train, y_train, steps = 200, batch_size = 32);

INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Restoring parameters from /tmp/tmp2ywxwlmv/model.ckpt-200
INFO:tensorflow:Saving checkpoints for 201 into /tmp/tmp2ywxwlmv/model.ckpt.
INFO:tensorflow:loss = 0.0720581, step = 201
INFO:tensorflow:global_step/sec: 258.84
INFO:tensorflow:loss = 0.05086443, step = 301 (0.386 sec)
INFO:tensorflow:Saving checkpoints for 400 into /tmp/tmp2ywxwlmv/model.ckpt.
INFO:tensorflow:Loss for final step: 0.039798416.


## Getting a list of the predictions

In [9]:
note_predictions = list(classifier.predict(X_test));

Instructions for updating:
Please switch to predict_classes, or set `outputs` argument.
Instructions for updating:
Estimator is decoupled from Scikit Learn interface by moving into
separate class SKCompat. Arguments x, y and batch_size are only
available in the SKCompat class, Estimator will only accept input_fn.
Example conversion:
  est = Estimator(...) -> est = SKCompat(Estimator(...))
INFO:tensorflow:Restoring parameters from /tmp/tmp2ywxwlmv/model.ckpt-400


## Reportings and Accuracy

In [11]:
from sklearn.metrics import confusion_matrix, classification_report
print("Confusion Matrix: ","\n"*2,confusion_matrix(y_test,note_predictions),"\n"*2)
print("Classification Report: ","\n"*2,classification_report(y_test,note_predictions))

Confusion Matrix:  

 [[13  0  0]
 [ 0 15  0]
 [ 0  1 16]] 


Classification Report:  

              precision    recall  f1-score   support

          0       1.00      1.00      1.00        13
          1       0.94      1.00      0.97        15
          2       1.00      0.94      0.97        17

avg / total       0.98      0.98      0.98        45



## Conclusion:
**With a quite fewer lines of code we were able to run TensorFlow via SciKitearn's '.contrib' method and rather expectedly we were able get a very accurate preductions averaging 98% for each class of iris. **